# [Validation] Data Analytics Test - iFood
###### [By Israel Mendes](https://github.com/israelmendez232)

This notebook were generated by [Databricks](https://databricks.com/). The idea is to replicate more possible of the enviroment of the Data Analytics from iFood.

## 1. Data Integrity
To make sure the data received is "making" sense.

In [0]:
df = spark.read.csv("/FileStore/tables/ml_project1_data-1.csv", header="true", inferSchema="true") 

display(df).head()

ID Year_Birth Education Marital_Status Income Kidhome Teenhome Dt_Customer Recency MntWines MntFruits MntMeatProducts MntFishProducts MntSweetProducts MntGoldProds NumDealsPurchases NumWebPurchases NumCatalogPurchases NumStorePurchases NumWebVisitsMonth AcceptedCmp3 AcceptedCmp4 AcceptedCmp5 AcceptedCmp1 AcceptedCmp2 Complain Z_CostContact Z_Revenue Response 5524 1957 Graduation Single 58138 0 0 2012-09-04 58 635 88 546 172 88 88 3 8 10 4 7 0 0 0 0 0 0 3 11 1 2174 1954 Graduation Single 46344 1 1 2014-03-08 38 11 1 6 2 1 6 2 1 1 2 5 0 0 0 0 0 0 3 11 0 4141 1965 Graduation Together 71613 0 0 2013-08-21 26 426 49 127 111 21 42 1 8 2 10 4 0 0 0 0 0 0 3 11 0 6182 1984 Graduation Together 26646 1 0 2014-02-10 26 11 4 20 10 3 5 2 2 0 4 6 0 0 0 0 0 0 3 11 0 5324 1981 PhD Married 58293 1 0 2014-01-19 94 173 43 118 46 27 15 5 5 3 6 5 0 0 0 0 0 0 3 11 0 7446 1967 Master Together 62513 0 1 2013-09-09 16 520 42 98 0 42 14 2 6 4 10 6 0 0 0 0 0 0 3 11 0 965 1971 Graduation Divorced 55635 0 1 2012-11-13 34 235 65 164 50 49 27 4 7 3 7 6 0 0 0 0 0 0 3 11 0 6177 1985 PhD Married 33454 1 0 2013-05-08 32 76 10 56 3 1 23 2 4 0 4 8 0 0 0 0 0 0 3 11 0 4855 1974 PhD Together 30351 1 0 2013-06-06 19 14 0 24 3 3 2 1 3 0 2 9 0 0 0 0 0 0 3 11 1 5899 1950 PhD Together 5648 1 1 2014-03-13 68 28 0 6 1 1 13 1 1 0 0 20 1 0 0 0 0 0 3 11 0 1994 1983 Graduation Married null 1 0 2013-11-15 11 5 5 6 0 2 1 1 1 0 2 7 0 0 0 0 0 0 3 11 0 387 1976 Basic Married 7500 0 0 2012-11-13 59 6 16 11 11 1 16 1 2 0 3 8 0 0 0 0 0 0 3 11 0 2125 1959 Graduation Divorced 63033 0 0 2013-11-15 82 194 61 480 225 112 30 1 3 4 8 2 0 0 0 0 0 0 3 11 0 8180 1952 Master Divorced 59354 1 1 2013-11-15 53 233 2 53 3 5 14 3 6 1 5 6 0 0 0 0 0 0 3 11 0 2569 1987 Graduation Married 17323 0 0 2012-10-10 38 3 14 17 6 1 5 1 1 0 3 8 0 0 0 0 0 0 3 11 0 2114 1946 PhD Single 82800 0 0 2012-11-24 23 1006 22 115 59 68 45 1 7 6 12 3 0 0 1 1 0 0 3 11 1 9736 1980 Graduation Married 41850 1 1 2012-12-24 51 53 5 19 2 13 4 3 3 0 3 8 0 0 0 0 0 0 3 11 0 4939 1946 Graduation Together 37760 0 0 2012-08-31 20 84 5 38 150 12 28 2 4 1 6 7 0 0 0 0 0 0 3 11 0 6565 1949 Master Married 76995 0 1 2013-03-28 91 1012 80 498 0 16 176 2 11 4 9 5 0 0 0 1 0 0 3 11 0 2278 1985 2n Cycle Single 33812 1 0 2012-11-03 86 4 17 19 30 24 39 2 2 1 3 6 0 0 0 0 0 0 3 11 0 9360 1982 Graduation Married 37040 0 0 2012-08-08 41 86 2 73 69 38 48 1 4 2 5 8 0 0 0 0 0 0 3 11 0 5376 1979 Graduation Married 2447 1 0 2013-01-06 42 1 1 1725 1 1 1 15 0 28 0 1 0 0 0 0 0 0 3 11 0 1993 1949 PhD Married 58607 0 1 2012-12-23 63 867 0 86 0 0 19 3 2 3 9 8 0 1 0 0 0 0 3 11 0 4047 1954 PhD Married 65324 0 1 2014-01-11 0 384 0 102 21 32 5 3 6 2 9 4 0 0 0 0 0 0 3 11 0 1409 1951 Graduation Together 40689 0 1 2013-03-18 69 270 3 27 39 6 99 7 7 1 5 8 0 0 0 0 0 0 3 11 0 7892 1969 Graduation Single 18589 0 0 2013-01-02 89 6 4 25 15 12 13 2 2 1 3 7 0 0 0 0 0 0 3 11 0 2404 1976 Graduation Married 53359 1 1 2013-05-27 4 173 4 30 3 6 41 4 5 1 4 7 0 0 0 0 0 0 3 11 0 5255 1986 Graduation Single null 1 0 2013-02-20 19 5 1 3 3 263 362 0 27 0 0 1 0 0 0 0 0 0 3 11 0 9422 1989 Graduation Married 38360 1 0 2013-05-31 26 36 2 42 20 21 10 2 2 1 4 3 0 0 0 0 0 0 3 11 0 1966 1965 PhD Married 84618 0 0 2013-11-22 96 684 100 801 21 66 0 1 6 9 10 2 0 0 1 0 0 0 3 11 0 6864 1989 Master Divorced 10979 0 0 2014-05-22 34 8 4 10 2 2 4 2 3 0 3 5 0 0 0 0 0 0 3 11 0 3033 1963 Master Together 38620 0 0 2013-05-11 56 112 17 44 34 22 89 1 2 5 3 3 0 0 0 0 0 0 3 11 0 5710 1970 Graduation Together 40548 0 1 2012-10-10 31 110 0 5 2 0 3 2 2 1 4 5 0 1 0 0 0 0 3 11 0 7373 1952 PhD Divorced 46610 0 2 2012-10-29 8 96 12 96 33 22 43 6 4 1 6 6 0 0 0 0 0 0 3 11 1 8755 1946 Master Married 68657 0 0 2013-02-20 4 482 34 471 119 68 22 1 3 5 9 7 0 0 0 0 0 0 3 11 0 10738 1951 Master Single 49389 1 1 2013-08-29 55 40 0 19 2 1 3 1 2 0 3 7 0 0 0 0 0 0 3 11 0 4339 1970 PhD Married 67353 0 1 2013-12-31 37 702 17 151 0 8 35 5 5 6 12 2 0 0 0 0 0 0 3 11 0 10755 1976 2n Cycle Married 23718 1 0 2013-09-02 76 6 3 14 15 7 36 3 3 1 

In [0]:
df.describe("ID", "Year_Birth", "Education", "Marital_Status", "Income", "Kidhome", "Teenhome", "Dt_Customer", "Recency", "MntWines", "MntFruits", "MntMeatProducts", "MntFishProducts", "MntSweetProducts", "MntGoldProds", "NumDealsPurchases", "NumWebPurchases").show()

+-------+------------------+------------------+---------+--------------+------------------+-------------------+------------------+-----------+-----------------+------------------+------------------+-----------------+------------------+-----------------+----------------+-----------------+------------------+
summary| ID| Year_Birth|Education|Marital_Status| Income| Kidhome| Teenhome|Dt_Customer| Recency| MntWines| MntFruits| MntMeatProducts| MntFishProducts| MntSweetProducts| MntGoldProds|NumDealsPurchases| NumWebPurchases|
+-------+------------------+------------------+---------+--------------+------------------+-------------------+------------------+-----------+-----------------+------------------+------------------+-----------------+------------------+-----------------+----------------+-----------------+------------------+
 count| 2240| 2240| 2240| 2240| 2216| 2240| 2240| 2240| 2240| 2240| 2240| 2240| 2240| 2240| 2240| 2240| 2240|
 mean| 5592.159821428571|1968.8058035714287| null| null| 52247.25135379061|0.44419642857142855| 0.50625| null| 49.109375| 303.9357142857143|26.302232142857143| 166.95| 37.52544642857143|27.06294642857143| 44.021875| 2.325| 4.084821428571429|
 stddev|3246.6621975643416|11.984069456885827| null| null|25173.076660901414| 0.5383980977345935|0.5445382307698761| null|28.96245280837821|336.59739260537197|39.773433764578584|225.7153725117536|54.628979402878095|41.28049848785491|52.1674389149973| 1.93223750085598|2.7787141473881007|
 min| 0| 1893| 2n Cycle| Absurd| 1730| 0| 0| 2012-07-30| 0| 0| 0| 0| 0| 0| 0| 0| 0|
 max| 11191| 1996| PhD| YOLO| 666666| 2| 2| 2014-06-29| 99| 1493| 199| 1725| 259| 263| 362| 15| 27|
+-------+------------------+------------------+---------+--------------+------------------+-------------------+------------------+-----------+-----------------+------------------+------------------+-----------------+------------------+-----------------+----------------+-----------------+------------------+

In [0]:
df.describe("NumCatalogPurchases", "NumStorePurchases", "NumWebVisitsMonth", "AcceptedCmp3", "AcceptedCmp4", "AcceptedCmp5", "AcceptedCmp1", "AcceptedCmp2", "Complain", "Z_CostContact", "Z_Revenue", "Response").show()

+-------+-------------------+-----------------+-----------------+-------------------+-------------------+-------------------+-------------------+--------------------+-------------------+-------------+---------+-------------------+
summary|NumCatalogPurchases|NumStorePurchases|NumWebVisitsMonth| AcceptedCmp3| AcceptedCmp4| AcceptedCmp5| AcceptedCmp1| AcceptedCmp2| Complain|Z_CostContact|Z_Revenue| Response|
+-------+-------------------+-----------------+-----------------+-------------------+-------------------+-------------------+-------------------+--------------------+-------------------+-------------+---------+-------------------+
 count| 2240| 2240| 2240| 2240| 2240| 2240| 2240| 2240| 2240| 2240| 2240| 2240|
 mean| 2.6620535714285714|5.790178571428571|5.316517857142857|0.07276785714285715|0.07455357142857143|0.07276785714285715|0.06428571428571428|0.013392857142857142| 0.009375| 3.0| 11.0|0.14910714285714285|
 stddev| 2.9231006555397463|3.250958145674428| 2.42664500954729| 0.2598130699218951|0.26272828485356176| 0.2598130699218953|0.24531597433401406| 0.11497560625548417|0.09639116794449748| 0.0| 0.0|0.35627358640847934|
 min| 0| 0| 0| 0| 0| 0| 0| 0| 0| 3| 11| 0|
 max| 28| 13| 20| 1| 1| 1| 1| 1| 1| 3| 11| 1|
+-------+-------------------+-----------------+-----------------+-------------------+-------------------+-------------------+-------------------+--------------------+-------------------+-------------+---------+-------------------+

From the **validation** is possible to notice a few problems:

- `Year_Birth:` Is possible to notice the lowest age in the dataset is 1893 and the oldest is 1996. I would check it again the source, because the range is to waid to make sense. Specially while doublechecking with `Dt_Customer`;
- `Income:` Is missing 24 responses. While the whole dataset is 2.240 rows long, the NULL values compose of 1.07% of the table. Which makes sense to drop that rows to avoid further problems with the EDA;
- `Marital_Status:` A few responses are way outside of the expected, such as _Absurd_ and _YOLO_;
- `Z_CostContact and Z_Revenue:` Has - exactly - the same element, the first one is _3_ and the second is _11_.

In [0]:
# Create a view or table

temp_table_name = "ml_project1_data"

df.createOrReplaceTempView(temp_table_name)

In [0]:
%sql

/* Query the created temp table in a SQL cell */

SELECT Income
FROM `ml_project1_data`
WHERE Income IS NULL
ORDER BY Income

Income null null null null null null null null null null null null null null null null null null null null null null null null

Now it's time to understand better the distinct values for the categorical variables.

In [0]:
%sql

SELECT DISTINCT Education
FROM `ml_project1_data`
ORDER BY Education

Education 2n Cycle Basic Graduation Master PhD

In [0]:
%sql

SELECT DISTINCT Marital_Status
FROM `ml_project1_data`
ORDER BY Marital_Status

Marital_Status Absurd Alone Divorced Married Single Together Widow YOLO

To make sure that there isn't any duplicate in the dataset.

In [0]:
%sql

SELECT COUNT(DISTINCT ID)
FROM `ml_project1_data`

count(DISTINCT ID) 2240

Excluding the NULL or NaN values

In [0]:
df_transformation = df.na.drop()
display(df_transformation).head()

ID Year_Birth Education Marital_Status Income Kidhome Teenhome Dt_Customer Recency MntWines MntFruits MntMeatProducts MntFishProducts MntSweetProducts MntGoldProds NumDealsPurchases NumWebPurchases NumCatalogPurchases NumStorePurchases NumWebVisitsMonth AcceptedCmp3 AcceptedCmp4 AcceptedCmp5 AcceptedCmp1 AcceptedCmp2 Complain Z_CostContact Z_Revenue Response 5524 1957 Graduation Single 58138 0 0 2012-09-04 58 635 88 546 172 88 88 3 8 10 4 7 0 0 0 0 0 0 3 11 1 2174 1954 Graduation Single 46344 1 1 2014-03-08 38 11 1 6 2 1 6 2 1 1 2 5 0 0 0 0 0 0 3 11 0 4141 1965 Graduation Together 71613 0 0 2013-08-21 26 426 49 127 111 21 42 1 8 2 10 4 0 0 0 0 0 0 3 11 0 6182 1984 Graduation Together 26646 1 0 2014-02-10 26 11 4 20 10 3 5 2 2 0 4 6 0 0 0 0 0 0 3 11 0 5324 1981 PhD Married 58293 1 0 2014-01-19 94 173 43 118 46 27 15 5 5 3 6 5 0 0 0 0 0 0 3 11 0 7446 1967 Master Together 62513 0 1 2013-09-09 16 520 42 98 0 42 14 2 6 4 10 6 0 0 0 0 0 0 3 11 0 965 1971 Graduation Divorced 55635 0 1 2012-11-13 34 235 65 164 50 49 27 4 7 3 7 6 0 0 0 0 0 0 3 11 0 6177 1985 PhD Married 33454 1 0 2013-05-08 32 76 10 56 3 1 23 2 4 0 4 8 0 0 0 0 0 0 3 11 0 4855 1974 PhD Together 30351 1 0 2013-06-06 19 14 0 24 3 3 2 1 3 0 2 9 0 0 0 0 0 0 3 11 1 5899 1950 PhD Together 5648 1 1 2014-03-13 68 28 0 6 1 1 13 1 1 0 0 20 1 0 0 0 0 0 3 11 0 387 1976 Basic Married 7500 0 0 2012-11-13 59 6 16 11 11 1 16 1 2 0 3 8 0 0 0 0 0 0 3 11 0 2125 1959 Graduation Divorced 63033 0 0 2013-11-15 82 194 61 480 225 112 30 1 3 4 8 2 0 0 0 0 0 0 3 11 0 8180 1952 Master Divorced 59354 1 1 2013-11-15 53 233 2 53 3 5 14 3 6 1 5 6 0 0 0 0 0 0 3 11 0 2569 1987 Graduation Married 17323 0 0 2012-10-10 38 3 14 17 6 1 5 1 1 0 3 8 0 0 0 0 0 0 3 11 0 2114 1946 PhD Single 82800 0 0 2012-11-24 23 1006 22 115 59 68 45 1 7 6 12 3 0 0 1 1 0 0 3 11 1 9736 1980 Graduation Married 41850 1 1 2012-12-24 51 53 5 19 2 13 4 3 3 0 3 8 0 0 0 0 0 0 3 11 0 4939 1946 Graduation Together 37760 0 0 2012-08-31 20 84 5 38 150 12 28 2 4 1 6 7 0 0 0 0 0 0 3 11 0 6565 1949 Master Married 76995 0 1 2013-03-28 91 1012 80 498 0 16 176 2 11 4 9 5 0 0 0 1 0 0 3 11 0 2278 1985 2n Cycle Single 33812 1 0 2012-11-03 86 4 17 19 30 24 39 2 2 1 3 6 0 0 0 0 0 0 3 11 0 9360 1982 Graduation Married 37040 0 0 2012-08-08 41 86 2 73 69 38 48 1 4 2 5 8 0 0 0 0 0 0 3 11 0 5376 1979 Graduation Married 2447 1 0 2013-01-06 42 1 1 1725 1 1 1 15 0 28 0 1 0 0 0 0 0 0 3 11 0 1993 1949 PhD Married 58607 0 1 2012-12-23 63 867 0 86 0 0 19 3 2 3 9 8 0 1 0 0 0 0 3 11 0 4047 1954 PhD Married 65324 0 1 2014-01-11 0 384 0 102 21 32 5 3 6 2 9 4 0 0 0 0 0 0 3 11 0 1409 1951 Graduation Together 40689 0 1 2013-03-18 69 270 3 27 39 6 99 7 7 1 5 8 0 0 0 0 0 0 3 11 0 7892 1969 Graduation Single 18589 0 0 2013-01-02 89 6 4 25 15 12 13 2 2 1 3 7 0 0 0 0 0 0 3 11 0 2404 1976 Graduation Married 53359 1 1 2013-05-27 4 173 4 30 3 6 41 4 5 1 4 7 0 0 0 0 0 0 3 11 0 9422 1989 Graduation Married 38360 1 0 2013-05-31 26 36 2 42 20 21 10 2 2 1 4 3 0 0 0 0 0 0 3 11 0 1966 1965 PhD Married 84618 0 0 2013-11-22 96 684 100 801 21 66 0 1 6 9 10 2 0 0 1 0 0 0 3 11 0 6864 1989 Master Divorced 10979 0 0 2014-05-22 34 8 4 10 2 2 4 2 3 0 3 5 0 0 0 0 0 0 3 11 0 3033 1963 Master Together 38620 0 0 2013-05-11 56 112 17 44 34 22 89 1 2 5 3 3 0 0 0 0 0 0 3 11 0 5710 1970 Graduation Together 40548 0 1 2012-10-10 31 110 0 5 2 0 3 2 2 1 4 5 0 1 0 0 0 0 3 11 0 7373 1952 PhD Divorced 46610 0 2 2012-10-29 8 96 12 96 33 22 43 6 4 1 6 6 0 0 0 0 0 0 3 11 1 8755 1946 Master Married 68657 0 0 2013-02-20 4 482 34 471 119 68 22 1 3 5 9 7 0 0 0 0 0 0 3 11 0 10738 1951 Master Single 49389 1 1 2013-08-29 55 40 0 19 2 1 3 1 2 0 3 7 0 0 0 0 0 0 3 11 0 4339 1970 PhD Married 67353 0 1 2013-12-31 37 702 17 151 0 8 35 5 5 6 12 2 0 0 0 0 0 0 3 11 0 10755 1976 2n Cycle Married 23718 1 0 2013-09-02 76 6 3 14 15 7 36 3 3 1 2 7 0 0 0 0 0 0 3 11 0 8595 1973 Graduation Widow 42429 0 1 2014-02-11 99 55 0 6 2 0 4 2 1 1 3 5 0 0 0 0 0 0 3 11 0 2968 1943 PhD Divorced 48948 0 0 2013-02-01 53 437 8 206 160 49 42 2 7 10 

Counting if there is any NULL in the dataframe yet.

In [0]:
from pyspark.sql.functions import col, sum

df_transformation.select(*(sum(col(c).isNull().cast("int")).alias(c) for c in df_transformation.columns)).show()

+---+----------+---------+--------------+------+-------+--------+-----------+-------+--------+---------+---------------+---------------+----------------+------------+-----------------+---------------+-------------------+-----------------+-----------------+------------+------------+------------+------------+------------+--------+-------------+---------+--------+
 ID|Year_Birth|Education|Marital_Status|Income|Kidhome|Teenhome|Dt_Customer|Recency|MntWines|MntFruits|MntMeatProducts|MntFishProducts|MntSweetProducts|MntGoldProds|NumDealsPurchases|NumWebPurchases|NumCatalogPurchases|NumStorePurchases|NumWebVisitsMonth|AcceptedCmp3|AcceptedCmp4|AcceptedCmp5|AcceptedCmp1|AcceptedCmp2|Complain|Z_CostContact|Z_Revenue|Response|
+---+----------+---------+--------------+------+-------+--------+-----------+-------+--------+---------+---------------+---------------+----------------+------------+-----------------+---------------+-------------------+-----------------+-----------------+------------+------------+------------+------------+------------+--------+-------------+---------+--------+
 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0|
+---+----------+---------+--------------+------+-------+--------+-----------+-------+--------+---------+---------------+---------------+----------------+------------+-----------------+---------------+-------------------+-----------------+-----------------+------------+------------+------------+------------+------------+--------+-------------+---------+--------+

### Outliers
Checking if the data doesn't have any outliers.

In [0]:
df.shape

--------------------------------------------------------------------------- 
 AttributeError Traceback (most recent call last)
 <command-449765189887898> in <module> 
 ----> 1 df . shape

 /databricks/spark/python/pyspark/sql/dataframe.py in __getattr__ (self, name) 
 1417 if name not in self . columns : 
 1418 raise AttributeError(
 -> 1419 "'%s' object has no attribute '%s'" % (self.__class__.__name__, name))
 1420 jc = self . _jdf . apply ( name ) 
 1421 return Column ( jc ) 

 AttributeError : 'DataFrame' object has no attribute 'shape'

In [0]:
Q1 = df.quantile(0.25)
Q3 = df.quantile(0.75)
IQR  = Q3 - Q1

--------------------------------------------------------------------------- 
 AttributeError Traceback (most recent call last)
 <command-449765189887899> in <module> 
 ----> 1 Q1 = df . quantile ( 0.25 ) 
 2 Q3 = df . quantile ( 0.75 ) 
 3 IQR = Q3 - Q1

 /databricks/spark/python/pyspark/sql/dataframe.py in __getattr__ (self, name) 
 1417 if name not in self . columns : 
 1418 raise AttributeError(
 -> 1419 "'%s' object has no attribute '%s'" % (self.__class__.__name__, name))
 1420 jc = self . _jdf . apply ( name ) 
 1421 return Column ( jc ) 

 AttributeError : 'DataFrame' object has no attribute 'quantile'